File: 10-data-structures/02-linked-list/01-linked-list-implementation-.js
/**
 * @class Node
 * @description Represents a single node in the linked list.
 *
 * Each Node holds a value and a reference (pointer) to the next Node in the sequence.
 */
class Node {
  /**
   * @constructor
   * @param {*} value The data to store in the node. Can be of any type.
   */
  constructor(value) {
    this.value = value; // The data stored in the node
    this.next = null; // Reference to the next node in the list
  }
}

/**
 * @class LinkedList
 * @description Implements a singly linked list data structure.
 *
 * This class provides methods for common linked list operations such as
 * appending, prepending, deleting nodes, and printing the list.
 */
class LinkedList {
  /**
   * @constructor
   * @param {*} value The value for the first node of the list.
   */
  constructor(value) {
    // Create the first node and set it as both head and tail
    this.head = new Node(value);
    this.tail = this.head;
    this.length = 1;
  }

  /**
   * @method append
   * @description Adds a new node with the given value to the end of the list.
   * @param {*} value The value to append.
   * @returns {LinkedList} The current LinkedList instance for chaining.
   */
  append(value) {
    const newNode = new Node(value);
    // If the list is empty (shouldn't happen with initial constructor, but good for robustness)
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode; // Link the current tail to the new node
      this.tail = newNode; // Update tail to the new node
    }
    this.length++;
    return this; // Allow method chaining
  }

  /**
   * @method prepend
   * @description Adds a new node with the given value to the beginning of the list.
   * @param {*} value The value to prepend.
   * @returns {LinkedList} The current LinkedList instance for chaining.
   */
  prepend(value) {
    const newNode = new Node(value);
    // If the list is empty
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head; // Link the new node to the current head
      this.head = newNode; // Update head to the new node
    }
    this.length++;
    return this; // Allow method chaining
  }

  /**
   * @method printList
   * @description Prints all the values in the linked list from head to tail to the console.
   *
   * Displays the list in the format: value1 --> value2 --> ... --> null
   */
  printList(head = this.head) {
    let temp = this.head; // Start from the head
    let listString = "";
    while (temp !== null) {
      listString += `${temp.value} --> `; // Append current node's value
      temp = temp.next; // Move to the next node
    }
    listString += "null"; // Indicate the end of the list
    console.log(listString);
  }

  /**
   * @method deleteLast
   * @description Deletes the last node from the linked list.
   *
   * Handles cases for empty lists, single-node lists, and multi-node lists.
   * @returns {*} The value of the deleted node, or null if the list was empty.
   */
  deleteLast() {
    if (!this.head) {
      // Check if the list is empty
      console.warn("Warning: Cannot delete from an empty list.");
      return null;
    }

    let deletedValue = null;

    if (this.head === this.tail) {
      // If there's only one node in the list
      deletedValue = this.head.value;
      this.head = null; // Reset head and tail
      this.tail = null;
    } else {
      let temp = this.head;
      // Traverse until 'temp' is the node *before* the tail
      while (temp.next !== this.tail) {
        temp = temp.next;
      }
      deletedValue = this.tail.value;
      this.tail = temp; // Update tail to the new last node
      this.tail.next = null; // The new tail points to null
    }
    this.length--; // Decrement the list length
    return deletedValue;
  }

  /**
   * @method deleteFirst
   * @description Deletes the first node (head) from the linked list.
   *
   * Handles cases for empty lists and single-node lists.
   * @returns {*} The value of the deleted node, or null if the list was empty.
   */
  deleteFirst() {
    if (!this.head) {
      // Check if the list is empty
      console.warn("Warning: Cannot delete from an empty list.");
      return null;
    }

    const deletedValue = this.head.value;
    const temp = this.head; // Store the current head for potential return value

    if (this.head.next === null) {
      // If there's only one node (head is also tail)
      this.head = null; // Reset head and tail
      this.tail = null;
    } else {
      this.head = this.head.next; // Move head to the next node
    }
    // In JavaScript, temp (the old head) will be garbage collected if no other references exist.
    // We don't need to 'delete' it explicitly.
    this.length--; // Decrement the list length
    return deletedValue;
  }

  /**
   * @method getLength
   * @description Gets the current number of nodes in the linked list.
   * @returns {number} The length of the linked list.
   */
  getLength() {
    return this.length;
  }

  /**
   * @method isEmpty
   * @description Checks if the linked list is empty.
   * @returns {boolean} True if the list is empty, false otherwise.
   */
  isEmpty() {
    return this.length === 0;
  }
  /**
   * @method reverse
   * @description Reverse the linked list
   * @returns {newHead} True if the list is empty, false otherwise.
   */

  reverseLinkedListBruteForce() {
    const st = [];
    let temp = this.head;
    while (temp) {
      st.push(temp.value);
      temp = temp.next;
    }
    temp = this.head;
    while (temp) {
      temp.value = st.pop();
      temp = temp.next;
    }
  }

  reverseLinkedListOptimal() {
    let temp = this.head;
    let prev = null;
    while (temp !== null) {
      let front = temp.next; // Store next node
      temp.next = prev; // Reverse the link
      prev = temp; // Move prev forward
      temp = front; // Move temp forward
    }
    this.tail = this.head; // Update tail to old head
    this.head = prev; // Update head to new head
    return this.head;
  }
}

// --- Main execution block for demonstration ---
console.log("--- Initializing List ---");
const my_list = new LinkedList(10);

my_list.printList(); // Output: 10 --> null
console.log("Length:", my_list.getLength()); // Output: Length: 1

console.log("\n--- Appending Values ---");
my_list.append(20).append(30); // Chaining append calls
my_list.printList(); // Output: 10 --> 20 --> 30 --> null
console.log("Length:", my_list.getLength()); // Output: Length: 3

console.log("\n--- Prepending Value ---");
my_list.prepend(5);
my_list.printList(); // Output: 5 --> 10 --> 20 --> 30 --> null
console.log("Length:", my_list.getLength()); // Output: Length: 4

console.log("\n--- Deleting First Node ---");
const deletedFirst = my_list.deleteFirst();
console.log("Deleted first node value:", deletedFirst); // Output: Deleted first node value: 5
my_list.printList(); // Output: 10 --> 20 --> 30 --> null
console.log("Length:", my_list.getLength()); // Output: Length: 3

console.log("\n--- Deleting Last Node ---");
const deletedLast = my_list.deleteLast();
console.log("Deleted last node value:", deletedLast); // Output: Deleted last node value: 30
my_list.printList(); // Output: 10 --> 20 --> null
console.log("Length:", my_list.getLength()); // Output: Length: 2

console.log("\n--- Deleting Remaining Nodes ---");
my_list.deleteFirst(); // Delete 10
my_list.printList(); // Output: 20 --> null
my_list.deleteLast(); // Delete 20
my_list.printList(); // Output: null
console.log("Length:", my_list.getLength()); // Output: Length: 0
console.log("Is empty?", my_list.isEmpty()); // Output: Is empty? true

console.log("\n--- Attempting to Delete from Empty List ---");
my_list.deleteFirst(); // Output: Warning: Cannot delete from an empty list.
my_list.deleteLast(); // Output: Warning: Cannot delete from an empty list.

my_list.append(50);
my_list.append(60);
my_list.append(70);
my_list.append(80);
my_list.printList();
// my_list.reverseLinkedListBruteForce();
const newHead = my_list.reverseLinkedListOptimal();
// console.log(newHead);
my_list.printList();

File: 10-data-structures/02-linked-list/01-linked-list-implementation.cpp
#include <iostream>  // Required for input/output operations, like std::cout

/**
 * @brief Represents a single node in the linked list.
 *
 * Each Node holds an integer value and a pointer to the next Node in the sequence.
 */
class Node
{
   public:
    int value;   // The data stored in the node
    Node* next;  // Pointer to the next node in the list

    /**
     * @brief Constructs a new Node with the given value.
     * @param val The integer value to store in the node.
     */
    Node(int val) : value(val), next(nullptr) {}
};

/**
 * @brief Implements a singly linked list data structure.
 *
 * This class provides methods for common linked list operations such as
 * appending, prepending, deleting nodes, and printing the list.
 */
class LinkedList
{
   private:
    Node* head;  // Pointer to the first node in the list
    Node* tail;  // Pointer to the last node in the list
    int length;  // Current number of nodes in the list

   public:
    /**
     * @brief Constructs a new LinkedList with an initial node.
     * @param value The value for the first node of the list.
     */
    LinkedList(int value)
    {
        // Create the first node and set it as both head and tail
        head = new Node(value);
        tail = head;
        length = 1;
    }

    /**
     * @brief Destroys the LinkedList, deallocating all nodes.
     *
     * This destructor iterates through the list and frees the memory
     * occupied by each node to prevent memory leaks.
     */
    ~LinkedList()
    {
        Node* current = head;
        while (current != nullptr)
        {
            Node* nextNode = current->next;  // Store next node before deleting current
            delete current;                  // Deallocate current node
            current = nextNode;              // Move to the next node
        }
        head = nullptr;  // Ensure head and tail are null after deletion
        tail = nullptr;
        length = 0;  // Reset length
    }

    /**
     * @brief Adds a new node with the given value to the end of the list.
     * @param value The value to append.
     */
    void append(int value)
    {
        Node* newNode = new Node(value);
        // If the list is empty, the new node becomes both head and tail
        if (head == nullptr)
        {
            head = newNode;
            tail = newNode;
        }
        else
        {
            tail->next = newNode;  // Link the current tail to the new node
            tail = newNode;        // Update tail to the new node
        }
        length++;
    }

    /**
     * @brief Adds a new node with the given value to the beginning of the list.
     * @param value The value to prepend.
     */
    void prepend(int value)
    {
        Node* newNode = new Node(value);
        // If the list is empty, the new node becomes both head and tail
        if (head == nullptr)
        {
            head = newNode;
            tail = newNode;
        }
        else
        {
            newNode->next = head;  // Link the new node to the current head
            head = newNode;        // Update head to the new node
        }
        length++;
    }

    /**
     * @brief Prints all the values in the linked list from head to tail.
     *
     * Displays the list in the format: value1 --> value2 --> ... --> nullptr
     */
    void printList() const
    {
        Node* temp = head;  // Start from the head
        while (temp != nullptr)
        {
            std::cout << temp->value << " --> ";  // Print current node's value
            temp = temp->next;                    // Move to the next node
        }
        std::cout << "nullptr" << std::endl;  // Indicate the end of the list
    }

    /**
     * @brief Deletes the last node from the linked list.
     *
     * Handles cases for empty lists, single-node lists, and multi-node lists.
     */
    void deleteLast()
    {
        if (head == nullptr)
        {  // Check if the list is empty
            std::cout << "Error: Cannot delete from an empty list." << std::endl;
            return;
        }

        if (head == tail)
        {                    // If there's only one node in the list
            delete head;     // Deallocate the single node
            head = nullptr;  // Reset head and tail
            tail = nullptr;
        }
        else
        {
            Node* temp = head;
            // Traverse until 'temp' is the node *before* the tail
            while (temp->next != tail)
            {
                temp = temp->next;
            }
            delete tail;           // Deallocate the tail node
            tail = temp;           // Update tail to the new last node
            tail->next = nullptr;  // The new tail points to nullptr
        }
        length--;  // Decrement the list length
    }

    /**
     * @brief Deletes the first node (head) from the linked list.
     *
     * Handles cases for empty lists and single-node lists.
     */
    void deleteFirst()
    {
        if (head == nullptr)
        {  // Check if the list is empty
            std::cout << "Error: Cannot delete from an empty list." << std::endl;
            return;
        }

        Node* temp = head;  // Store the current head
        if (head->next == nullptr)
        {                    // If there's only one node (head is also tail)
            head = nullptr;  // Reset head and tail
            tail = nullptr;
        }
        else
        {
            head = head->next;  // Move head to the next node
        }
        delete temp;  // Deallocate the original head node
        length--;     // Decrement the list length
    }

    /**
     * @brief Gets the current number of nodes in the linked list.
     * @return The length of the linked list.
     */
    int getLength() const { return length; }

    bool insert(int value, int index)
    {
        if (head == nullptr)
        {
            prepend(value);
            return true;
        }
        if (index == length - 1)
        {
            append(value);
            return true;
        }
        Node* newNode = new Node(value);
        Node* prev = head;
        for (int i = 0; i < index; i++)
        {
            prev = prev->next;
        }
        newNode->next = prev->next;
        prev->next = newNode;
        length++;
        return true;
    }

    bool deletePosition(int index)
    {
        if (head == nullptr)
        {
            std::cout << "Empty List" << std::endl;
            return false;
        }

        if (index < 0 || index >= length)
        {
            std::cout << "Index out of bounds" << std::endl;
            return false;
        }

        Node* toDelete = nullptr;

        if (index == 0)
        {
            toDelete = head;
            head = head->next;
        }
        else
        {
            Node* prev = head;
            for (int i = 0; i < index - 1; i++)
            {
                prev = prev->next;
            }

            toDelete = prev->next;
            prev->next = toDelete->next;
        }

        delete toDelete;
        length--;
        return true;
    }

    // 1->2->3->4->5
    // 5->2->3->4->1
    void reverseLinkedList()
    {
        Node* temp = head;
        head = tail;
        tail = temp;
        Node* before = nullptr;
        Node* after = temp->next;
        for (int i = 0; i < length; i++)
        {
            after = temp->next;
            temp->next = before;
            before=temp;
            temp = after;
            // temp = after;
        }
    }
};

// --- Main function for demonstration ---
int main()
{
    // Create a new linked list with an initial value of 10
    LinkedList my_list(10);
    my_list.printList();  // Output: 10 --> nullptr

    // Append a value to the list
    my_list.append(20);
    my_list.append(30);
    my_list.printList();  // Output: 10 --> 20 --> 30 --> nullptr

    // Prepend a value to the list
    my_list.prepend(5);
    my_list.printList();  // Output: 5 --> 10 --> 20 --> 30 --> nullptr

    // Delete the first node
    my_list.deleteFirst();
    my_list.printList();  // Output: 10 --> 20 --> 30 --> nullptr

    // Delete the last node
    my_list.deleteLast();
    my_list.printList();  // Output: 10 --> 20 --> nullptr

    // Demonstrate deleting all nodes
    my_list.deleteFirst();  // Delete 10
    my_list.printList();    // Output: 20 --> nullptr
    my_list.deleteLast();   // Delete 20
    my_list.printList();    // Output: nullptr

    // Attempt to delete from an empty list
    my_list.deleteFirst();  // Output: Error: Cannot delete from an empty list.
    LinkedList* myll = new LinkedList(10);
    myll->append(20);
    myll->append(30);
    myll->append(40);
    myll->append(50);

    myll->printList();
    myll->reverseLinkedList();
    myll->printList();
    return 0;  // Indicate successful execution
}